[ACP-99] Specify a standard `ACP99Manager` abstract contract (#170)
avalanche-foundation/ACPs: ACPs/99-validatorsetmanager-contract/README.md

> 

Reviewer: for some reason, I can't make suggestions on Line 17:
The P-Chain allows a Subnet to specify a "validator manager" if it chooses to convert to an L1 using `ConvertSubnetToL1Tx`. 
Outside of any direct reference to the `SubnetID` param, this ACP should only reference any management of L1 validators sets and Avalanche L1 networks.

Author: Updated this language to clarify.

On each validator set change, the P-Chain is willing to sign an `AddressedCall` to notify any on-chain program tracking the validator set. On-chain programs must be able to interpret this message, so they can trigger the appropriate action. The 2 kinds of `AddressedCall`s [defined in ACP-77](https://github.com/avalanche-foundation/ACPs/tree/main/ACPs/77-reinventing-subnets#p-chain-warp-message-payloads) are `L1ValidatorRegistrationMessage` and `L1ValidatorWeightMessage`.

> Given these assumptions and the fact that most of the active blockchains on Avalanche mainnet are EVM-based, we propose `ACP99Manager` as the standard Solidity contract specification that can:

Reviewer: ```suggestion
Given these assumptions and the fact that most of the active blockchains on Avalanche Mainnet are EVM-based, we propose `ACP99Manager` as the standard Solidity contract specification that can:
```

> 1. Hold relevant information about the current Subnet validator set

Reviewer: ```suggestion
1. Hold relevant information about the current L1 validator set
```

2. Send validator set updates to the P-Chain by generating `AdressedCall`s defined in ACP-77
3. Correctly update the validator set by interpreting notification messages received from the P-Chain

> 4. Be easily integrated into applications implementing various security models (e.g. Proof-of-Stake).

Author: Updated the language

Having an audited and open-source reference implementation freely available will contribute to lowering the cost of launching L1s on Avalanche.

> Once deployed, the `ACP99Manager` implementation contract will be used as the `Address` in the [`ConvertSubnetToL1Tx`](https://github.com/avalanche-foundation/ACPs/tree/main/ACPs/77-reinventing-subnets#convertsubnettol1tx).

Reviewer: ```suggestion
Once deployed, the `ACP99Manager` implementation contract can be used as the `Address` in the [`ConvertSubnetToL1Tx`](https://github.com/avalanche-foundation/ACPs/tree/main/ACPs/77-reinventing-subnets#convertsubnettol1tx).
```

}

/**
 * @notice Contains the active state of a Validation
 * @notice status is the validation status
 * @notice nodeID is The NodeID of the validator
 * @notice startingWeight is the weight of the validator at the time of registration
 * @notice messageNonce is the current weight update nonce
 * @notice weight is the current weight of the validator
 * @notice startTime The start time of the validation

>  * @notice endTime The end time of the validation

Author: `@param` only pertains to [functions, events, and errors](https://docs.soliditylang.org/en/latest/natspec-format.html#tags), so I went with the more generic `@notice` here. Happy to change it to `@dev` if you think that would be more appropriate, but it seems NatSpec is not all that useful for documenting structs.

>  * @notice endTime The end time of the validation

Reviewer 2: Even tho the `@param` tag is officially supported only for functions, events, and errors, the documentation generation still looks better IMO. But I am okay with using `@notice` if you don't feel comfortable using `@param`.

Using `@notice`:
![image](https://github.com/user-attachments/assets/37bc4e8a-38fd-4f37-853c-6eff61807138)

Using `@param`:
![image](https://github.com/user-attachments/assets/fff30471-2ecf-473b-b9a0-577c553180d7)

>  * @notice status is the validation status
>  * @notice nodeID is The NodeID of the validator
>  * @notice startingWeight is the weight of the validator at the time of registration
>  * @notice messageNonce is the current weight update nonce
>  * @notice weight is the current weight of the validator
>  * @notice startTime The start time of the validation
>  * @notice endTime The end time of the validation

Reviewer 2: If sticking to using `@notice`:
```suggestion
 * @notice status is the validation status
 * @notice nodeID is the NodeID of the validator
 * @notice startingWeight is the weight of the validator at the time of registration
 * @notice messageNonce is the current weight update nonce
 * @notice weight is the current weight of the validator
 * @notice startTime is the start time of the validation
 * @notice endTime is the end time of the validation
```

>  * @notice endTime The end time of the validation

Author: Nice, I hadn't seen it rendered and assumed @param wouldn't be supported in this context. Happy to switch over.

 */
struct Validation {
    ValidationStatus status;
    bytes nodeID;
    uint64 startingWeight;
    uint64 messageNonce;
    uint64 weight;
    uint64 startTime;
    uint64 endTime;
}
```

#### About `Validation`s

A `Validation` represents the continuous time frame during which a node is part of the validator set and can be composed of multiple periods. A new period starts every time the validator weight changes during the same validation.

> Each `Validation` is identified by its `validationID` which is the SHA256 of the Payload of the `AddressedCall` in the `RegisterL1ValidatorTx` adding the validator to the Subnet's validator set, as defined in ACP-77.

Reviewer: ```suggestion
Each `Validation` is identified by its `validationID` which is the SHA256 of the Payload of the `AddressedCall` in the `RegisterL1ValidatorTx` adding the validator to the L1's validator set, as defined in ACP-77.
```

 */
abstract contract ACP99Manager {
    /// @notice Emitted when an initial validator is registered
    event RegisteredInitialValidator(
        bytes32 indexed nodeID, bytes32 indexed validationID, uint64 weight
    );
    /// @notice Emitted when a validator registration to the L1 is initiated
    event InitiatedValidatorRegistration(
        bytes32 indexed nodeID,
        bytes32 indexed validationID,
        bytes32 registrationMessageID,
        uint64 registrationExpiry,
        uint64 weight
    );
    /// @notice Emitted when a validator registration to the L1 is completed
    event CompletedValidatorRegistration(
        bytes32 indexed nodeID, bytes32 indexed validationID, uint64 weight
    );
    /// @notice Emitted when a validator weight update is initiated
    event InitiatedValidatorWeightUpdate(
        bytes32 indexed nodeID,
        bytes32 indexed validationID,
        bytes32 weightUpdateMessageID,
        uint64 weight
    );
    /// @notice Emitted when a validator weight update is completed
    event CompletedValidatorWeightUpdate(
        bytes32 indexed nodeID, bytes32 indexed validationID, uint64 nonce, uint64 weight
    );

>     /// @notice Returns the ID of the Subnet tied to this manager

Reviewer: ```suggestion
    /// @notice Returns the SubnetID of the L1 tied to this manager
```

>     function subnetID() virtual public view returns (bytes32);

Author: Technically, `subnetID` is correct here. It's not as explicit as it could be, but note that the ACP-77 message specifications use `subnetID` to identify the converted L1. Also note that the `subnetID` is the tx hash of the `CreateSubnetTx`, and therefore does not change when the Subnet is converted to an L1. As such, changing the name of the identifier of a static value depending on the Subnet/L1 status would be confusing in its own right. Eventually, once Subnets are fully deprecated in favor of L1s, it will make more sense to revisit this terminology.

This could be much better documented (and in fact, some contracts in the `icm-contracts` repo linked in this PR need to be changed back to `subnetID`), but ACP-77 is `Active` and therefore considered locked.

>     function subnetID() virtual public view returns (bytes32);

Reviewer 2: Makes sense. I only suggested changing here because v1.0.0 of `ValidatorManager` uses `_l1ID`:

https://github.com/ava-labs/icm-contracts/blob/046ca6d83c07c4cffecf61c06f4156b8b08725cc/contracts/validator-manager/ValidatorManager.sol#L40

>     /// @notice Returns the ID of the Subnet tied to this manager
>     function subnetID() virtual public view returns (bytes32);

Reviewer: I'd suggest: 
```suggestion
    /// @notice Returns the SubnetID of the L1 tied to this manager
    function subnetID() virtual public view returns (bytes32);
```
or 
```suggestion
    /// @notice Returns the SubnetID of the L1 tied to this manager
    function l1ID() virtual public view returns (bytes32);
```

}

/**
 * @notice Contains the active state of a Validator.
 * @param status The validator status.
 * @param nodeID The NodeID of the validator.
 * @param startingWeight The weight of the validator at the time of registration.
 * @param messageNonce The current weight update nonce.
 * @param weight The current weight of the validator.
 * @param startTime The start time of the validator.
 * @param endTime The end time of the validator.
 */
struct Validator {
    ValidatorStatus status;
    bytes nodeID;
    uint64 startingWeight;
    uint64 messageNonce;
    uint64 weight;
    uint64 startTime;
    uint64 endTime;
}
```

#### About `Validator`s

A `Validator` represents the continuous time frame during which a node is part of the validator set and can be composed of multiple periods. A new period starts every time the validator's weight changes.

> Each `Validator` is identified by its `validationID` which is the SHA256 of the Payload of the `AddressedCall` in the `RegisterL1ValidatorTx` adding the validator to the L1's validator set, as defined in ACP-77.

Reviewer: ```suggestion
Each `Validator` is identified by its `validationID`. If a validator was added as part of the initial set of continuous dynamic fee paying validators, its `validationID` is the SHA256 hash of the `ConvertSubnetToL1Tx` ID and the index of the initial validator. If a validator was added to the L1's validator set post-conversion, its `validationID` is the SHA256 of the payload of the `AddressedCall` in the `RegisterL1ValidatorTx` used to add it, as defined in ACP-77. 
```

}

/**
 * @notice Contains the active state of a Validator.
 * @param status The validator status.
 * @param nodeID The NodeID of the validator.
 * @param startingWeight The weight of the validator at the time of registration.
 * @param sentNonce The current weight update nonce sent by the manager.
 * @param receivedNonce The highest nonce received from the P-Chain
 * @param weight The current weight of the validator.
 * @param startTime The start time of the validator.
 * @param endTime The end time of the validator.
 */
struct Validator {
    ValidatorStatus status;
    bytes nodeID;
    uint64 startingWeight;
    uint64 sentNonce;
    uint64 receivedNonce;
    uint64 weight;
    uint64 startTime;
    uint64 endTime;
}
```

#### About `Validator`s

> A `Validator` represents the continuous time frame during which a node is part of the validator set and can be composed of multiple periods. A new period starts every time the validator's weight changes.

Reviewer 2: Does it still make sense to talk about validation periods? (since they were removed from the spec)
```suggestion
A `Validator` represents the continuous time frame during which a node is part of the validator set.
```

Author: Agreed, validation periods delineated by weight changes are out of scope.

/*
 * @title ACP99Manager
 * @notice The ACP99Manager interface represents the functionality for sovereign L1
 * validator management, as specified in ACP-77.
 */
abstract contract ACP99Manager {
    /// @notice Emitted when an initial validator is registered.
    event RegisteredInitialValidator(

>         bytes32 indexed validationID, bytes nodeID, uint64 weight

Reviewer 2: What's the rationale for removing `nodeID` from the indexed topics? IMO it is valuable to filter those events by NodeID to find a node's validations history.

Reviewer: looks like it is still there, just swapped place and typing with `validationID`.

```
bytes32 indexed validationID, **bytes nodeID**, uint64 weight
```

Reviewer 2: Yes, but it is not indexed anymore.

Reviewer: ah i understand now, and agree that it would be helpful for tracking metrics per node

Reviewer 2: Makes sense. I was unaware of this limitation in Solidity.

In my first implementation of ACP-99, I used a `bytes32` to store the NodeID as it is capable of storing both the current 20-byte format and the ACP-20 32-byte format. Do you reckon that there would be other format > 32 bytes in the future?

Author: After discussing with @geoff-vball, we agree that the best compromise is to accept `bytes nodeID` in functions, and emit `bytes20 nodeID` in events. Dynamic `bytes` in functions simplifies the upgrade path if the `nodeID` format ever changes. Since the implementation must convert from dynamic to fixed bytes to emit the event, we might as well emit `bytes20` to exactly reflect the current format. Future `nodeID` format changes would require contract validation changes anyway, so the event can be updated at that time.

Finally, since we do have to pay gas to convert `bytes` -> `bytes20` (around 800 gas in my tests), I updated the events to only emit the `nodeID` once per `validationID`. The `validationID` is indexed in every event, so the corresponding `nodeID` should be easily recoverable even if it's not emitted in the every event.

     * @param disableOwner The disable owner of the validator.
     * @param weight The weight of the node on the L1.
     * @return validationID The ID of the registered validator.
     */
    function _initiateValidatorRegistration(
        bytes memory nodeID,
        bytes memory blsPublicKey,
        uint64 registrationExpiry,
        PChainOwner memory remainingBalanceOwner,
        PChainOwner memory disableOwner,
        uint64 weight
    ) virtual internal returns (bytes32 validationID);

    /**
     * @notice Completes the validator registration process by returning an acknowledgement of the registration of a
     * validationID from the P-Chain. The validator should not be considered active until this method is successfully called.
     *
     * Emits a {CompletedValidatorRegistration} event on success.
     *
     * @param messageIndex The index of the L1ValidatorRegistrationMessage to be received providing the acknowledgement.
     * @return validationID The ID of the registered validator.
     */
    function completeValidatorRegistration(
        uint32 messageIndex
    ) virtual public returns  (bytes32 validationID);

    /**

>     * @notice Initiates validator removal by issuing a L1ValidatorWeightMessage with the weight set to zero.

Reviewer 2: ```suggestion
    * @notice Initiates validator removal by issuing an L1ValidatorWeightMessage with the weight set to zero.
```

    * The validator should be considered inactive as soon as this function is called.
    *
    * Emits an {InitiatedValidatorRemoval} on success.
    *
    * @param validationID The ID of the validator to remove.
    */
    function _initiateValidatorRemoval(
        bytes32 validationID
    ) virtual internal;

    /**

>     * @notice Completes validator removal by consuming an RegisterL1ValidatorMessage from the P-Chain acknowledging

Reviewer 2: ```suggestion
    * @notice Completes validator removal by consuming a RegisterL1ValidatorMessage from the P-Chain acknowledging
```

    * that the validator has been removed.
    *
    * Emits a {CompletedValidatorRemoval} on success.
    *
    * @param messageIndex The index of the RegisterL1ValidatorMessage.
    */
    function completeValidatorRemoval(
        uint32 messageIndex
    ) virtual public returns (bytes32 validationID);

    /**

>      * @notice Initiates validator weight update by issuing a L1ValidatorWeightMessage with a nonzero weight.

Reviewer 2: ```suggestion
     * @notice Initiates a validator weight update by issuing an L1ValidatorWeightMessage with a nonzero weight.
```

     * The validator weight change should not have any effect until completeValidatorWeightUpdate is successfully called.
     *
     * Emits an {InitiatedValidatorWeightUpdate} event on success.
     *
     * @param validationID The ID of the validator to modify.
     * @param weight The new weight of the validator.
     * @return nonce The validator nonce associated with the weight change.
     * @return messageID The ID of the L1ValidatorWeightMessage used to update the validator's weight.
     */
    function _initiateValidatorWeightUpdate(
        bytes32 validationID,
        uint64 weight
    ) virtual internal returns (uint64 nonce, bytes32 messageID);

    /**

>      * @notice Completes the validator weight update process by consuming a L1ValidatorWeightMessage from the P-Chain

Reviewer 2: ```suggestion
     * @notice Completes the validator weight update process by consuming an L1ValidatorWeightMessage from the P-Chain
```

#### About `DisableL1ValidatorTx`

This transaction allows the `DisableOwner` of a validator to disable it directly from the P-Chain to claim the unspent `Balance` linked to the validation of a failed L1. Therefore it is not meant to be called in the `Manager` contract.

## Backwards Compatibility

`ACP99Manager` is a reference specification. As such, it doesn't have any impact on the current behavior of the Avalanche protocol.

## Reference Implementation

A reference implementation will be provided in Ava Labs' [ICM Contracts](https://github.com/ava-labs/icm-contracts/tree/main/contracts/validator-manager) repository. This reference implementation will need to be updated to conform to `ACP99Manager` before this ACP may be marked `Implementable`.

### Example Integrations

`ACP99Manager` is designed to be easily incorporated into any architecture. Two example integrations are included in this ACP, each of which uses a different architecture.

> #### Two-contract Design

Reviewer 2: "Multi-contract Design" would be more generic (see the [BalancerValidatorManager](https://github.com/suzaku-network/suzaku-contracts-library/blob/main/src/contracts/ValidatorManager/BalancerValidatorManager.sol) that can have multiple security modules managing registrations/weight changes)
```suggestion
#### Multi-contract Design
```

> The two-contract design consists of a contract that implements `ACP99Manager`, and a separate "security module" contract that implements a single staking security model, such as PoS or PoA. Each `ACP99Manager` implementation contract is associated with one "security module" that is the only contract allowed to call the `ACP99Manager` functions that initiate validator set changes (`initiateValidatorRegistration`, and `initiateValidatorWeightUpdate`). Every time a validator is added/removed or a weight change is initiated, the `ACP99Manager` implementation will in turn call the corresponding function of the "security module" (`handleValidatorRegistration` or `handleValidatorWeightChange`). We recommend that the "security module" reference an immutable `ACP99Manager` contract address for security reasons.

Reviewer 2: I wouldn't qualify PoA as "staking".
```suggestion
The multi-contract design consists of a contract that implements `ACP99Manager`, and separate "security module" contracts that implement security models, such as PoS or PoA. Each `ACP99Manager` implementation contract is associated with one or more "security modules" that are the only contracts allowed to call the `ACP99Manager` functions that initiate validator set changes (`initiateValidatorRegistration`, and `initiateValidatorWeightUpdate`). Every time a validator is added/removed or a weight change is initiated, the `ACP99Manager` implementation will, in turn, call the corresponding function of the "security module" (`handleValidatorRegistration` or `handleValidatorWeightChange`). We recommend that the "security modules" reference an immutable `ACP99Manager` contract address for security reasons.
```

Author: > Each `ACP99Manager` implementation contract is associated with one or more "security modules"

Curious if there is a specific use case in mind that would use more than one security module. Or are you clarifying that the security module need not be a single contract? Either way, implementation of such a design and enforcement by the `ValidatorManager` thereof is outside the scope of this ACP.

Reviewer 2: The use case we have in mind is allowing a smoother transition between security models (than what is available by having a single security module). See the [BalancerValidatorManager](https://github.com/suzaku-network/suzaku-contracts-library/blob/main/src/contracts/ValidatorManager/BalancerValidatorManager.sol) that allows to set a max total weight per security module. An example would be giving 800 max weight to a PoA security module and 200 max weight to a PoS security module so start decentralizing the network and then allocating more and more weight on the PoS side.

Author: Very cool idea, thanks for linking.

It is up to the "security module" to decide what action to take when a validator is added/removed or a weight change is confirmed by the P-Chain. Such actions could be starting the withdrawal period and allocating rewards in a PoS L1.

```mermaid
---

> title: ACP-99 Two-Contract Architecture with a PoA Security Module

Reviewer 2: ```suggestion
title: ACP-99 Multi-contract Architecture with a PoA Security Module
```

"Security modules" could implement PoS, Liquid PoS, etc. The specification of such smart contracts is out of the scope of this ACP.

> A work in progress implementation is available in the [Suzaku Contracts Library](https://github.com/suzaku-network/suzaku-contracts-library/blob/acp-99-implementation/README.md#acp99-contracts-library) repository. It will be updated until this ACP is considered `Implementable` based on the outcome of the discussion.

Reviewer: @Nuttymoon this link to suzaku is still returning a 404

```mermaid
---
title: ACP-99 Single-Contract Architecture implementing PoA
---
  classDiagram
  class ACP99Manager {
    initializeValidatorSet
    initiateValidatorRegistration
    completeValidatorRegistration
    initiateValidatorWeightUpdate
    completeValidatorWeightUpdate
  }
  <<abstract>> ACP99Manager

  class ValidatorManager {
    completeValidatorRegistration
  }
  <<abstract>> ValidatorManager
  class PoAValidatorManager {
    initiateValidatorRegistration
    initiateEndValidation
    completeEndValidation
  }

  ACP99Manager <|--ValidatorManager
  ValidatorManager <|-- PoAValidatorManager
```

This implementation is available in Ava Labs' [ICM Contracts Repository](https://github.com/ava-labs/icm-contracts/tree/main/contracts/validator-manager).

## Security Considerations

The audit process of `ACP99Manager` and reference implementations is of the utmost importance for the future of the Avalanche ecosystem as most L1s would rely upon it to secure their L1.

## Open Questions

### Is there an interest to keep historical information about the validator set on the manager chain?

> It's undefined if `getValidation` should return allow historical information about ended validations. Should `ACP99Manager` enforce that validation information be kept in the contract's state indefinitely? Note that validator performance (uptime) is _not_ specified in the `ACP99Manager` interface, as it may not be relevant to some applications (e.g. PoA). Historical uptime would be a useful metric to query, but it may be more appropriately left to implementations to enforce.

Reviewer 2: ```suggestion
It's undefined if `getValidation` should return historical information about ended validations. Should `ACP99Manager` enforce that validation information be kept in the contract's state indefinitely? Note that validator performance (uptime) is _not_ specified in the `ACP99Manager` interface, as it may not be relevant to some applications (e.g. PoA). Historical uptime would be a useful metric to query, but it may be more appropriately left to implementations to enforce.
```

If we don't require `ACP99Manager` implementations to keep track of historical validations, this information will still be available in archive nodes and offchain tools (e.g. explorers).

### Should `ACP99Manager` include a churn control mechanism?

The Ava Labs [implementation](https://github.com/ava-labs/icm-contracts/blob/main/contracts/validator-manager/ValidatorManager.sol) of the `ValidatorManager` contract includes a churn control mechanism that prevents too much weight from being added or removed from the validator set in a short amount of time. Excessive churn can cause consensus failures, so it may be appropriate to require that churn tracking is implemented in some capacity.

## Acknowledgments

> 

Reviewer: @cam-schultz you definitely deserve a spot on this list!

```mermaid
---
title: ACP-99 Single-Contract Architecture implementing PoA
---
  classDiagram
  class ACP99Manager {
    initializeValidatorSet
    initiateValidatorRegistration
    completeValidatorRegistration
    initiateValidatorWeightUpdate
    completeValidatorWeightUpdate
  }
  <<abstract>> ACP99Manager

  class ValidatorManager {
    completeValidatorRegistration
  }
  <<abstract>> ValidatorManager
  class PoAValidatorManager {
    initiateValidatorRegistration
    initiateEndValidation
    completeEndValidation
  }

  ACP99Manager <|--ValidatorManager
  ValidatorManager <|-- PoAValidatorManager
```

This implementation is available in Ava Labs' [ICM Contracts Repository](https://github.com/ava-labs/icm-contracts/tree/main/contracts/validator-manager).

## Security Considerations

The audit process of `ACP99Manager` and reference implementations is of the utmost importance for the future of the Avalanche ecosystem as most L1s would rely upon it to secure their L1.

## Open Questions

### Is there an interest to keep historical information about the validator set on the manager chain?

> It's undefined if `getValidator` should return historical information about historical validators. Should `ACP99Manager` enforce that validator information be kept in the contract's state indefinitely? Note that validator performance (uptime) is _not_ specified in the `ACP99Manager` interface, as it may not be relevant to some applications (e.g. PoA). Historical uptime would be a useful metric to query, but it may be more appropriately left to implementations to enforce.

Reviewer 3: Maybe "it's left to the implementor" instead of "it's undefined"?

Author: I've rewritten this whole section to be less awkwardly worded.
